File: 1917-maximum-average-pass-ratio/1917-maximum-average-pass-ratio.cpp

class Solution {
public:
    double maxAverageRatio(vector<vector<int>>& classes, int extraStudents) {
        int n = classes.size();
        auto compare = [&](int a, int b) {
            double q1 =
                (((double)classes[a][0] + 1) / ((double)classes[a][1] + 1)) -
                ((double)classes[a][0] / classes[a][1]);
            double q2 =
                (((double)classes[b][0] + 1) / ((double)classes[b][1] + 1)) -
                ((double)classes[b][0] / classes[b][1]);
            return q1 < q2;
        };

        priority_queue<int, vector<int>, decltype(compare)> pq(compare);

        for (int i = 0; i < n; i++) {
            pq.push(i);
        }

        for (int i = 0; i < extraStudents; i++) {
            int topIndex = pq.top();
            pq.pop();
            classes[topIndex][0]++;
            classes[topIndex][1]++;
            pq.push(topIndex);
        }

        double res = 0;
        for (int i = 0; i < n; i++) {
            res += ((double)classes[i][0] / classes[i][1] / n);
        }
        return res;
    }
};

File: 1917-maximum-average-pass-ratio/1917-maximum-average-pass-ratio.py
class Solution:
    def maxAverageRatio(self, classes: List[List[int]], extraStudents: int) -> float:
        res = 0
        pq = []
        def priority(idx: int) -> float:
            return -(((classes[idx][0] + 1) / (classes[idx][1] + 1)) - (classes[idx][0] / classes[idx][1]))
        for i in range(len(classes)):
            heapq.heappush(pq, (priority(i), i))
        for i in range(extraStudents):
            topIdx = heappop(pq)[1]
            classes[topIdx][0] += 1
            classes[topIdx][1] += 1
            heapq.heappush(pq, (priority(topIdx), topIdx))
        for passed, total in classes:
            res += (passed / total) / len(classes)
        return res
        